### PR TITLE
Initial topdown optimizations

### PR DESCRIPTION
**Pull request by tsandall**
This PR contains a few optimizations that I made to the topdown evaluation last week after doing some profiling. The individual optimization commits are fairly small and isolated.

Here's the output of running the scheduler benchmark on the last 6 commits:

```
[work](benchmark)torin:~/Development/Styra/work/src/github.com/open-policy-agent/opa$ for rev in $(git rev-list HEAD...52b6684a8c03efe51cd4e166b8b60e5197cb747d | head -n 10 | xargs); do git checkout $rev; make perf; done
Note: checking out '533784940a1e1304fa9e6db1e780cd4e2fda690c'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at 5337849... Optimization: Avoid slice copies in hotspots
go generate
go test -v -bench=. ./test/perf/.../
testing: warning: no tests to run
PASS
BenchmarkScheduler10Nodes-8	       1	2803496994 ns/op
ok  	github.com/open-policy-agent/opa/test/perf/scheduler	2.854s
Previous HEAD position was 5337849... Optimization: Avoid slice copies in hotspots
HEAD is now at f106040... Optimization: Avoid plug on indexing check
go generate
go test -v -bench=. ./test/perf/.../
testing: warning: no tests to run
PASS
BenchmarkScheduler10Nodes-8	       1	3111978150 ns/op
ok  	github.com/open-policy-agent/opa/test/perf/scheduler	3.158s
Previous HEAD position was f106040... Optimization: Avoid plug on indexing check
HEAD is now at b840758... Optimization: Remove copying of bindings
go generate
go test -v -bench=. ./test/perf/.../
testing: warning: no tests to run
PASS
BenchmarkScheduler10Nodes-8	       1	3258621573 ns/op
ok  	github.com/open-policy-agent/opa/test/perf/scheduler	3.305s
Previous HEAD position was b840758... Optimization: Remove copying of bindings
HEAD is now at 30e7066... Optimization: Defer unification until needed
go generate
go test -v -bench=. ./test/perf/.../
testing: warning: no tests to run
PASS
BenchmarkScheduler10Nodes-8	       1	3757969676 ns/op
ok  	github.com/open-policy-agent/opa/test/perf/scheduler	3.804s
Previous HEAD position was 30e7066... Optimization: Defer unification until needed
HEAD is now at d0cd968... Optimization: Improve string and var term hashing
go generate
go test -v -bench=. ./test/perf/.../
testing: warning: no tests to run
PASS
BenchmarkScheduler10Nodes-8	       1	3994262904 ns/op
ok  	github.com/open-policy-agent/opa/test/perf/scheduler	4.041s
Previous HEAD position was d0cd968... Optimization: Improve string and var term hashing
HEAD is now at d12d3a4... Vendoring: Add siphash and update dependencies
go generate
go test -v -bench=. ./test/perf/.../
testing: warning: no tests to run
PASS
BenchmarkScheduler10Nodes-8	       1	4915227223 ns/op
ok  	github.com/open-policy-agent/opa/test/perf/scheduler	4.962s
[work](detached*)torin:~/Development/Styra/work/src/github.com/open-policy-agent/opa$
```

### CODE REVIEW

File: ast/term.go [hunk 1]
```diff
@@ -329,9 +331,8 @@ func (str String) String() string {
 
 // Hash returns the hash code for the Value.
 func (str String) Hash() int {
-	h := fnv.New64a()
-	h.Write([]byte(str))
-	return int(h.Sum64())
+	h := siphash.Hash(0xdeadbeef, 0xdeadbeef, *(*[]byte)(unsafe.Pointer(&str)))
```
Comment (tim-styra) [inline ast/term.go]: Hashing a constant is a red flag.  Maybe a comment explaining why this is ok.

Comment (tsandall) [inline ast/term.go]: Whoops, good catch! This should be using a random key.

**Comment by timothyhinrichs** [comment]
Github isn't letting me add this comment inline for some reason. It pertains to the Binding algorithm...

Trying to think through if this always gives us the right answers.

In this scheme we're (i) changing the existing dictionary and (ii) allocating a new struct to record the old value.  Every unbind changes the existing dictionary back to what it was.  Does every plug look up a value in a dictionary, or does it transitively close the individual bindings?  I'm wondering about an example like a sequence of bindings x:y, y:z, z:1.  What would the resulting binding look like..  x:1, y:1, z:1?  What would the Undo give us?  x:y, y:z, x:z  ?  If we're transitively closing the bindings on each lookup, then I'd think we'd have x:y, y:z, z:1 with an undo of x:y, y:z.  That seems like it should work, as long as the namespaces of the variables are always distinct.  Does this query work?  Query: p[z] = z.  Rule p[x] = y :- q[z] = (x,y).  Or does the z in the body of the rule get mistakenly bound to the z from the query?

An alternative, which seems like it will always give us the right answer, is where the datastructure is a linked-list of dictionaries.  On each bind we allocate another dictionary that contains the new bindings and push it onto the front of the list.  On each unbind, we pop the front element off the list.  On each lookup we traverse the list from front-to-back looking for a binding (recursively).  When binding a variable to a variable, we end up using a pointer to the variable we're binding to, which ensures the z's in the example above are different.

**Comment by timothyhinrichs** [comment]
Finished comments on this PR

**Comment by timothyhinrichs** [comment]
Forget about my binding comment.  Was confused about the roles of ctxt.Local versus ctxt. The reason is that there are 2 different ways that bindings get "undone" here.  One is the new one that backs up changes to a single ctxt.Local in the case of multiple rules with the same head.  The other one existed long before this patch and happens naturally when a recursive call to TopDown finishes (since then all the bindings made during the recursive call are "undone" by that ctxt being garbage-collected).   Anyway, LGTM once the constant-hash is fixed up.

**Pull request merged**